File: Src/Modules/conv.cs
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Runtime;
using IronPython.Runtime;
using System.Collections;

namespace NFox.Pycad
{

    public static class conv
    {

        public static List ToList(ResultBuffer rb)
        {
            List lst = new List();
            if (rb != null)
            {
                foreach (var value in rb)
                {
                    var slst = new List();
                    slst.append(value.TypeCode);
                    slst.append(value.Value);
                    lst.append(slst);
                }
            }
            return lst;
        }


        public static ResultBuffer ToBuffer(List lst)
        {
            ResultBuffer rb = new ResultBuffer();
            foreach (List slst in lst)
                rb.Add(new TypedValue((int)slst[0], slst[1]));
            return rb;
        }

        public static TypedValue[] ToTypedArray(List lst)
        {
            var arr = new TypedValue[lst.Count];
            for (int i = 0; i < lst.Count; i++)
            {
                var slst = lst[i] as List;
                arr[i] = new TypedValue((int)slst[0], slst[1]);
            }
            return arr;
        }

        public static SelectionFilter ToFilter(List lst)
        {
            return new SelectionFilter(ToTypedArray(lst));
        }

        #region LispData

        public static object FromLispData(ResultBuffer rb)
        {
            if (rb != null)
                return GetSubList(rb.GetEnumerator());
            return new List();
        }

        static object GetSubList(ResultBufferEnumerator it)
        {
            List lst = new List();
            while (it.MoveNext())
            {
                switch ((LispDataType)it.Current.TypeCode)
                {
                    case LispDataType.ListEnd:
                        return lst;
                    case LispDataType.DottedPair:
                        return new PythonTuple(lst);
                    case LispDataType.ListBegin:
                        lst.append(GetSubList(it));
                        break;
                    case LispDataType.T_atom:
                        lst.append(true);
                        break;
                    case LispDataType.Nil:
                        lst.append(false);
                        break;
                    default:
                        lst.append(it.Current.Value);
                        break;
                }
            }
            return lst;
        }

        public static object ToLispData(object obj)
        {

            ResultBuffer rb = new ResultBuffer();
            GetBuffer(obj, rb);
            var arr = rb.AsArray();
            switch (arr.Length)
            {
                case 0:
                    return CreateNewValue(LispDataType.Nil);
                case 1:
                    return arr[0];
                default:
                    return rb;
            }
        }

        static readonly TypedValue T = CreateNewValue(LispDataType.T_atom);
        static readonly TypedValue Nil = CreateNewValue(LispDataType.Nil);

        static TypedValue CreateNewValue(LispDataType code, object value)
        {
            return new TypedValue((int)code, value);
        }

        static TypedValue CreateNewValue(LispDataType code)
        {
            return new TypedValue((int)code);
        }

        static void Add(ResultBuffer rb, LispDataType code, object value)
        {
            rb.Add(CreateNewValue(code, value));
        }

        static void Add(ResultBuffer rb, LispDataType code)
        {
            rb.Add(CreateNewValue(code));
        }

        static void GetBuffer(object obj, ResultBuffer rb)
        {
            if (obj is TypedValue)
                rb.Add(obj);
            else if (obj is bool)
                Add(rb, (bool)obj ? LispDataType.T_atom : LispDataType.Nil);
            else if (obj is short)
                Add(rb, LispDataType.Int16, obj);
            else if (obj is int)
                Add(rb, LispDataType.Int32, obj);
            else if (obj is double)
                Add(rb, LispDataType.Double, obj);
            else if (obj is string)
                Add(rb, LispDataType.Text, obj);
            else if (obj is Point2d)
                Add(rb, LispDataType.Point2d, obj);
            else if (obj is Point3d)
                Add(rb, LispDataType.Point3d, obj);
            else if (obj is ObjectId)
                Add(rb, LispDataType.ObjectId, obj);
            else if (obj is SelectionSet)
                Add(rb, LispDataType.SelectionSet, obj);
            else if (obj is PythonTuple)
                GetBuffer((PythonTuple)obj, rb);
            else if (obj is IEnumerable)
                GetBuffer((IEnumerable)obj, rb);
        }

        static void GetBuffer(IEnumerable lst, ResultBuffer rb)
        {
            Add(rb, LispDataType.ListBegin);
            foreach (var obj in lst)
                GetBuffer(obj, rb);
            Add(rb, LispDataType.ListEnd);
        }

        static void GetBuffer(PythonTuple lst, ResultBuffer rb)
        {
            Add(rb, LispDataType.ListBegin);
            foreach (var obj in lst)
                GetBuffer(obj, rb);
            Add(rb, LispDataType.DottedPair);
        }

        #endregion

    }
}

File: Projects/Debug/pycad/sys/conv.py

def ToList(rb):
    try: return [[v.TypeCode, v.Value] for v in rb]
    except: return []

def ToBuffer(lst):
    try:
        t = tuple((_newvalue(*v) for v in lst))
        return acdb.ResultBuffer(t)
    except:
        pass

def ToTypedArray(lst):
    from System import Array
    return Array[acdb.TypedValue]([_newvalue(*v) for v in lst])

def ToFilter(lst):
    return aced.SelectionFilter(ToTypedArray(lst))

def FromLispData(rb):
    if rb: return _getlist(rb.GetEnumerator())
    return []

def _getlist(it):
    lst = []
    while it.MoveNext():
        from System import Enum
        code = Enum.ToObject(acrx.LispDataType, it.Current.TypeCode)
        if code == acrx.LispDataType.ListEnd:
            return lst
        elif code == acrx.LispDataType.DottedPair:
            return tuple(lst)
        elif code == acrx.LispDataType.ListBegin:
            lst.append(_getlist(it))
        elif code == acrx.LispDataType.T_atom:
            lst.append(True)
        elif code == acrx.LispDataType.Nil:
            lst.append(False)
        else:
            lst.append(it.Current.Value)
    return lst

def ToLispData(obj):
    lst = []
    _getbuffer(obj, lst)
    n = len(lst)
    if n == 0:
        return _newvalue(acrx.LispDataType.Nil)
    elif n == 1:
        return lst[0]
    return acdb.ResultBuffer(tuple(lst))

def _getbuffer(obj, lst):
    from System import Int16, Int32
    if isinstance(obj, acdb.TypedValue):
        lst.append(obj)
    elif isinstance(obj, (Int16, Int32, int)):
        lst.append(_newvalue(acrx.LispDataType.Int32, obj))
    elif obj == True:
        lst.append(_newvalue(acrx.LispDataType.T_atom))
    elif obj in (False, None):
        lst.append(_newvalue(crx.LispDataType.Nil))
    elif isinstance(obj, float):
        lst.append(_newvalue(acrx.LispDataType.Double, obj))
    elif isinstance(obj, str):
        lst.append(_newvalue(acrx.LispDataType.Text, obj))
    elif isinstance(obj, acge.Point2d):
        lst.append(_newvalue(acrx.LispDataType.Point2d, obj))
    elif isinstance(obj, acge.Point3d):
        lst.append(_newvalue(acrx.LispDataType.Point3d, obj))
    elif isinstance(obj, acdb.ObjectId):
        lst.append(_newvalue(acrx.LispDataType.ObjectId, obj))
    elif isinstance(obj, aced.SelectionSet):
        lst.append(_newvalue(acrx.LispDataType.SelectionSet, obj))
    else:
        from collections import Iterable
        if isinstance(obj, Iterable):
            lst.append(_newvalue(acrx.LispDataType.ListBegin))
            for o in obj:
                _getbuffer(o, lst)
            if isinstance(obj, tuple):
                lst.append(_newvalue(acrx.LispDataType.DottedPair))
            else:
                lst.append(_newvalue(acrx.LispDataType.ListEnd))

def _newvalue(code, value = -1):
    try:
        return acdb.TypedValue(code, value)
    except TypeError:
        return acdb.TypedValue(code.value__, value)

def _getlist2(it):
    lst = []
    while it.MoveNext():
        from System import Enum
        case = switcher(Enum.ToObject(acrx.LispDataType, it.Current.TypeCode))
        if case(acrx.LispDataType.ListEnd): return lst
        elif case(acrx.LispDataType.DottedPair): return tuple(lst)
        elif case(acrx.LispDataType.ListBegin): lst.append(_getlist2(it))
        elif case(acrx.LispDataType.T_atom): lst.append(True)
        elif case(acrx.LispDataType.Nil): lst.append(False)
        else: lst.append(it.Current.Value)
    return lst

def _getbuffer2(obj, lst):
    from System import Int16, Int32
    case = switcher(obj)
    if case[acdb.TypedValue]:
        lst.append(obj)
    elif case[Int16, Int32, int]:
        lst.append(_newvalue(acrx.LispDataType.Int32, obj))
    elif case(True):
        lst.append(_newvalue(acrx.LispDataType.T_atom))
    elif case(False, None):
        lst.append(_newvalue(crx.LispDataType.Nil))
    elif case[float]:
        lst.append(_newvalue(acrx.LispDataType.Double, obj))
    elif case[str]:
        lst.append(_newvalue(acrx.LispDataType.Text, obj))
    elif case[acge.Point2d]:
        lst.append(_newvalue(acrx.LispDataType.Point2d, obj))
    elif case[acge.Point3d]:
        lst.append(_newvalue(acrx.LispDataType.Point3d, obj))
    elif case[acdb.ObjectId]:
        lst.append(_newvalue(acrx.LispDataType.ObjectId, obj))
    elif case[aced.SelectionSet]:
        lst.append(_newvalue(acrx.LispDataType.SelectionSet, obj))
    else:
        from collections import Iterable
        if case[Iterable]:
            lst.append(_newvalue(acrx.LispDataType.ListBegin))
            for o in obj:
                _getbuffer2(o, lst)
            if case[tuple]:
                lst.append(_newvalue(acrx.LispDataType.DottedPair))
            else:
                lst.append(_newvalue(acrx.LispDataType.ListEnd))